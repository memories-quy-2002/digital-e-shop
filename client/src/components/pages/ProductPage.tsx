import { useEffect, useState } from "react";
import { BsStar, BsStarFill } from "react-icons/bs";
import axios from "../../api/axios";
import { useToast } from "../../context/ToastContext";
import "../../styles/ProductPage.scss";
import { Product } from "../../utils/interface";
import ratingStar from "../../utils/ratingStar";
import NavigationBar from "../common/NavigationBar";
import RecommendedProduct from "../common/RecommendedProduct";
import Layout from "../layout/Layout";
import NoPage from "./NoPage";
import { useAuth } from "../../context/AuthContext";
import { Helmet } from "react-helmet";
import { useLocation } from "react-router-dom";
interface relevantProductsItem {
    product_id: number;
    product_name: string;
}

type Wishlist = {
    id: number;
    product: Product;
};

type Review = {
    username: string;
    rating: number;
    reviewText: string;
    created_at: Date;
};

const ProductPage = () => {
    const location = useLocation();
    const url = new URLSearchParams(location.search);
    const { addToast } = useToast();
    const { uid } = useAuth();
    const productId = url.get("id");
    const pid = productId !== null ? parseInt(productId) : 0;
    const [products, setProducts] = useState<Product[]>([]);
    const [productDetail, setProductDetail] = useState<Product>({
        id: 0,
        name: "",
        category: "",
        brand: "",
        price: 0,
        sale_price: 0,
        rating: 0,
        reviews: 0,
        main_image: "",
        image_gallery: [],
        stock: 0,
        description: "",
        specifications: [],
    });
    const [relevantProducts, setRelevantProducts] = useState<Product[]>([]);
    const [ratingScore, setRatingScore] = useState<number>(0);
    const [reviewText, setReviewText] = useState<string>("");
    const [reviews, setReviews] = useState<Review[]>([]);
    const [wishlist, setWishlist] = useState<Wishlist[]>([]);
    const [toggle, setToggle] = useState<boolean>(false);
    const [quantity, setQuantity] = useState<number>(0);

    useEffect(() => {
        const fetchSingleProduct = async () => {
            try {
                const response = await axios.get(`/api/products/${pid}`);
                if (response.status === 200) {
                    setProductDetail(response.data.product);
                    // setQuantity(response.data.product.stock); // Cập nhật số lượng
                    console.log(response.data.msg);
                }
            } catch (err: any) {
                console.error(err);
            }
        };
        fetchSingleProduct();
        // Clean up the Blob URL after component unmount
        return () => {};
    }, [pid]);

    useEffect(() => {
        const fetchAllProducts = async () => {
            try {
                const response = await axios.get("/api/products");
                if (response.status === 200) {
                    setProducts(response.data.products);
                    console.log(response.data.msg);
                }
            } catch (err) {
                console.error(err);
            }
        };
        fetchAllProducts();
        return () => {};
    }, []);

    useEffect(() => {
        const fetchWishlist = async () => {
            try {
                const response = await axios.get(`/api/wishlist/${uid}`);
                if (response.status === 200) {
                    const newWishlist: Wishlist[] = response.data.wishlist.map((item: any) => {
                        const { id, product_id, ...productProps } = item;

                        return {
                            id,
                            product: {
                                id: product_id,
                                ...productProps,
                            },
                        };
                    });

                    setWishlist(newWishlist);
                    console.log(response.data.msg);
                }
            } catch (err) {
                console.error(err);
            }
        };
        fetchWishlist();
        return () => {};
    }, [uid]);

    useEffect(() => {
        const fetchRelevantProducts = async () => {
            try {
                const response = await axios.get(`/api/products/relevant/${pid}`);
                if (response.status === 200) {
                    if (response.data.relevantProducts.length === 0) {
                        setRelevantProducts(products);
                        return;
                    }
                    const newRelevantProducts: relevantProductsItem[] = response.data.relevantProducts;
                    const relevantProductsIds: number[] = newRelevantProducts.map((product) => product.product_id);

                    setRelevantProducts(
                        products
                            .filter((product) => relevantProductsIds.includes(product.id))
                            .sort((a, b) => {
                                const indexA = relevantProductsIds.indexOf(a.id);
                                const indexB = relevantProductsIds.indexOf(b.id);
                                return indexA - indexB;
                            })
                    );
                    console.log(response.data.msg);
                }
            } catch (err) {
                console.error(err);
            }
        };
        fetchRelevantProducts();

        return () => {};
    }, [products, pid]);

    useEffect(() => {
        const fetchReviews = async () => {
            try {
                const response = await axios.get(`/api/reviews/${productDetail.id}`);
                if (response.status === 200) {
                    setReviews(
                        response.data.reviews.map((review: any) => {
                            return {
                                username: review.username,
                                rating: review.rating,
                                reviewText: review.review_text,
                                created_at: review.created_at,
                            };
                        })
                    );
                    console.log(response.data.msg);
                }
            } catch (err) {
                console.error(err);
            }
        };
        fetchReviews();
        return () => {};
    }, [productDetail]);

    const handleDecrease = () => setQuantity((quantity) => quantity - 1);
    const handleIncrease = () => setQuantity((quantity) => quantity + 1);
    const handleStarClick = (rating: number) => {
        setRatingScore(rating);
    };

    const handleAddingCart = async (user_id: string, product: Product) => {
        if (uid === "") {
            addToast("Login required", "You need to login to use this feature.");
            return;
        } else {
            if (productDetail.stock === 0) {
                addToast("Out of stock", "The product is out of stock.");
                return;
            } else if (quantity === 0) {
                addToast("Invalid action", "The quantity should be non-zero to complete the action.");
                return;
            }
        }
        try {
            const response = await axios.post("/api/cart/", {
                uid: user_id,
                pid: product.id,
                quantity: quantity,
            });
            if (response.status === 200) {
                console.log(response.data.msg);
                addToast("Add cart item", "Product added to cart successfully.");
            }
        } catch (err) {
            throw err;
        }
    };

    const handleAddingWishlist = async (user_id: string, product_id: number) => {
        if (uid === "") {
            addToast("Login required", "You need to login to use this feature.");
            return;
        }
        try {
            if (wishlist.some((item) => item.product.id === product_id)) {
                const response = await axios.post(`/api/wishlist/delete/`, {
                    uid: user_id,
                    pid: product_id,
                });
                if (response.status === 200) {
                    console.log(response.data.msg);
                    setWishlist((list) => list.filter((item) => item.product.id !== product_id));
                    addToast("Remove wishlist item", "Item removed from wishlist successfully");
                }
            } else {
                const response = await axios.post("/api/wishlist/", {
                    uid: user_id,
                    pid: product_id,
                });
                if (response.status === 200) {
                    console.log(response.data.msg);
                    const newProduct = products.filter((product) => product.id === product_id)[0];
                    setWishlist((list) => [
                        ...list,
                        {
                            id: product_id,
                            product: newProduct,
                        },
                    ]);
                    addToast("Add wishlist item", "Item added to wishlist successfully");
                }
            }
        } catch (err) {
            console.error(err);
        }
    };

    const handleSubmitReview = async (uid: string, pid: number, rating: number, reviewText: string) => {
        if (ratingScore === 0) {
            addToast("Invalid rating", "The rating score should be from 1 to 5");
            return;
        }
        try {
            const response = await axios.post("/api/reviews/", {
                uid,
                pid,
                rating,
                reviewText,
            });

            if (response.status === 200) {
                console.log(response.data.msg);
                addToast("Submit review", "Review has been submitted successfully.");
                window.location.reload();
            }
        } catch (err) {
            throw err;
        }
    };

    if (pid <= 0) {
        return <NoPage />;
    }
    return (
        <Layout>
            <NavigationBar />
            <Helmet>
                <title>{productDetail.name}</title>
            </Helmet>
            <div className="product__container">
                <div className="product__container__detail">
                    <div className="product__container__detail__img">
                        {productDetail.main_image ? (
                            <img
                                src={`https://epgq6ejr4lgv8lec.public.blob.vercel-storage.com/uploads/${productDetail.main_image}.jpg`}
                                alt={productDetail.name}
                            />
                        ) : (
                            <img
                                src={require("../../assets/images/product_placeholder.jpg")}
                                alt={productDetail.name}
                            />
                        )}
                    </div>
                    <div className="product__container__detail__main">
                        <div className="product__container__detail__main__info">
                            <div className="product__container__detail__main__info__rating">
                                <div
                                    style={{
                                        display: "flex",
                                        flexDirection: "row",
                                        gap: "0.25rem",
                                    }}
                                >
                                    {ratingStar(productDetail.rating, "#FFCC4A", 24)}
                                </div>
                                <span
                                    style={{
                                        fontSize: "1.5rem",
                                    }}
                                >
                                    {productDetail.rating} ({productDetail.reviews})
                                </span>
                            </div>
                            <strong className="product__container__detail__main__info__name">
                                {productDetail.name}
                            </strong>

                            <div className="product__container__detail__main__info__price">
                                {productDetail.sale_price ? (
                                    <>
                                        <span className="product__container__detail__main__info__price-active">
                                            ${productDetail.sale_price}
                                        </span>{" "}
                                        <span className="product__container__detail__main__info__price-original">
                                            ${productDetail.price}
                                        </span>
                                    </>
                                ) : (
                                    <span className="product__container__detail__main__info__price-active">
                                        ${productDetail.price}
                                    </span>
                                )}
                            </div>
                            <div className="product__container__detail__main__info__other">
                                <span>Category: {productDetail.category}</span>
                                <span>Brand: {productDetail.brand}</span>
                                <span>
                                    Status:{" "}
                                    <span
                                        style={{
                                            color: "red",
                                            fontWeight: "bold",
                                        }}
                                    >
                                        {" "}
                                        {productDetail.stock > 0 ? "In stock" : "Out of stock"}
                                    </span>
                                </span>
                            </div>
                        </div>
                        <div className="product__container__detail__main__quantity">
                            <strong className="product__container__detail__main__quantity__title">Quantity</strong>
                            <div className="product__container__detail__main__quantity__input">
                                <button
                                    className="product__container__detail__main__quantity__input-minus"
                                    type="button"
                                    onClick={handleDecrease}
                                >
                                    -
                                </button>
                                <input
                                    type="number"
                                    name="quantity"
                                    id="quantity"
                                    value={quantity}
                                    onChange={(e: any) =>
                                        setProductDetail((detail) => {
                                            return {
                                                ...detail,
                                                stock: Number(e.target.value),
                                            };
                                        })
                                    }
                                />
                                <button
                                    className="product__container__detail__main__quantity__input-plus"
                                    type="button"
                                    onClick={handleIncrease}
                                >
                                    +
                                </button>
                            </div>
                        </div>
                        <div className="product__container__detail__main__button">
                            <button
                                className="product__container__detail__main__button-cart"
                                type="button"
                                onClick={() => {
                                    if (uid) {
                                        handleAddingCart(uid, productDetail);
                                    } else {
                                        addToast("Login required", "You need to login to use this feature.");
                                    }
                                }}
                            >
                                Add to cart
                            </button>
                            <button
                                className="product__container__detail__main__button-wishlist"
                                type="button"
                                onClick={() => {
                                    if (uid) {
                                        handleAddingWishlist(uid, productDetail.id);
                                    } else {
                                        addToast("Login required", "You need to login to use this feature.");
                                    }
                                }}
                            >
                                {wishlist.some((item) => item.product.id === pid)
                                    ? "Remove from wishlist"
                                    : "Add to wishlist"}
                            </button>
                        </div>
                    </div>
                </div>
                <div className="product__container__navigation">
                    <div className="product__container__navigation__btn">
                        <button type="button" className={!toggle ? "active" : ""} onClick={() => setToggle(false)}>
                            Description
                        </button>
                        <button type="button" className={toggle ? "active" : ""} onClick={() => setToggle(true)}>
                            Review ({reviews.length})
                        </button>
                        <div className="product__container__navigation__btn__empty"></div>
                    </div>
                </div>
                {toggle ? (
                    <div className="product__container__review">
                        {!uid ? (
                            <div className="product__container__review__container">
                                You need to login to use this feature
                            </div>
                        ) : (
                            <div className="product__container__review__container">
                                <div className="product__container__review__container__rating">
                                    <h5 className="product__container__review__container__rating__title">Rating</h5>
                                    <div className="product__container__review__container__rating__container">
                                        <div className="product__container__review__container__rating__container__star">
                                            {[1, 2, 3, 4, 5].map((rating) => (
                                                <span key={rating} onClick={() => handleStarClick(rating)}>
                                                    {rating <= ratingScore ? (
                                                        <BsStarFill size={18} color="#FFCC4A" />
                                                    ) : (
                                                        <BsStar size={18} color="#FFCC4A" />
                                                    )}
                                                </span>
                                            ))}
                                        </div>
                                        <div className="product__container__review__container__rating__container__score">
                                            <span>({ratingScore.toFixed(0)})</span>
                                        </div>
                                    </div>
                                </div>
                                <div className="product__container__review__container__text">
                                    <h5 className="product__container__review__container__text__title">Review</h5>
                                    <textarea
                                        className="product__container__review__container__text__area"
                                        name="review"
                                        id="review"
                                        value={reviewText}
                                        onChange={(event) => setReviewText(event.target.value)}
                                        rows={10}
                                        placeholder="Enter your review"
                                    ></textarea>
                                    <button
                                        className="product__container__review__container__text__button"
                                        type="submit"
                                        onClick={() =>
                                            handleSubmitReview(uid, productDetail.id, ratingScore, reviewText)
                                        }
                                    >
                                        Submit
                                    </button>
                                </div>
                                <div className="product__container__review__container__all">
                                    <h5 className="product__container__review__container__all__title">
                                        All reviews ({reviews.length})
                                    </h5>
                                    <div className="product__container__review__container__all__list">
                                        {reviews.map((review) => (
                                            <div className="product__container__review__container__all__list__item">
                                                <div className="product__container__review__container__all__list__item__info">
                                                    <strong>{review.username}</strong>
                                                    <span className="product__container__review__container__all__list__item__info__star">
                                                        {ratingStar(review.rating, "#FFCC4A", 18)}
                                                    </span>
                                                    <span>
                                                        {new Date(review.created_at).toLocaleDateString("en-GB")}
                                                    </span>
                                                </div>
                                                <div className="product__container__review__container__all__list__item__review">
                                                    <p>{review.reviewText}</p>
                                                </div>
                                            </div>
                                        ))}
                                    </div>
                                </div>
                            </div>
                        )}
                    </div>
                ) : (
                    <div className="product__container__description">
                        <h5>Description</h5>
                        <p>
                            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut
                            labore et dolore magna aliqua. Morbi tincidunt augue interdum velit euismod in pellentesque.
                            Nunc sed id semper risus in hendrerit gravida. Diam vel quam elementum pulvinar etiam non
                            quam lacus. Amet massa vitae tortor condimentum lacinia quis vel eros donec. Donec massa
                            sapien faucibus et. Dui ut ornare lectus sit. Quam pellentesque nec nam aliquam sem.
                            Ultricies integer quis auctor elit. Penatibus et magnis dis parturient montes nascetur
                            ridiculus mus mauris. Euismod in pellentesque massa placerat duis ultricies. Donec enim diam
                            vulputate ut pharetra sit amet. At tempor commodo ullamcorper a lacus vestibulum. Risus
                            pretium quam vulputate dignissim. Tincidunt tortor aliquam nulla facilisi cras. Accumsan
                            tortor posuere ac ut consequat semper viverra nam. Dictum sit amet justo donec. Ultricies
                            lacus sed turpis tincidunt id aliquet risus.
                            {productDetail.description}
                        </p>
                    </div>
                )}

                <RecommendedProduct pid={productDetail.id} relevantProducts={relevantProducts} />
            </div>
        </Layout>
    );
};

export default ProductPage;
