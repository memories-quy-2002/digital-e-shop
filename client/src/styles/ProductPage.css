.product__container {
  display: flex;
  flex-direction: column;
  padding: 1.5rem 5rem;
  gap: 4rem;
}
@media (min-width: 768px) {
  .product__container {
    padding: 3rem 10rem;
  }
}
.product__container__detail {
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 5rem;
  height: -moz-fit-content;
  height: fit-content;
}
@media (min-width: 768px) {
  .product__container__detail {
    flex-direction: row;
    gap: 1rem;
  }
}
.product__container__detail__img {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: center;
}
.product__container__detail__img img {
  width: 100%;
}
@media (min-width: 768px) {
  .product__container__detail__img img {
    width: 75%;
  }
}
.product__container__detail__main {
  flex: 1;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  gap: 2rem;
}
.product__container__detail__main__info {
  display: flex;
  flex-direction: column;
  gap: 0.75rem;
}
.product__container__detail__main__info__rating {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 2rem;
}
.product__container__detail__main__info__name {
  font-size: 1.5rem;
}
.product__container__detail__main__info__price {
  margin-top: 1rem;
  font-size: 1.75rem;
  font-weight: bold;
}
.product__container__detail__main__info__price-active {
  color: red;
}
.product__container__detail__main__info__price-original {
  margin-left: 1rem;
  color: gray;
  text-decoration: line-through;
}
.product__container__detail__main__info__other {
  margin-top: 1rem;
  font-size: 1.25rem;
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}
.product__container__detail__main__quantity {
  height: -moz-fit-content;
  height: fit-content;
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: 1rem;
}
.product__container__detail__main__quantity__title {
  font-size: 1.25rem;
}
.product__container__detail__main__quantity__input {
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid #ccc;
  border-radius: 8px;
  overflow: hidden;
  width: 9rem;
  height: 40px;
}
.product__container__detail__main__quantity__input input[type=number] {
  width: 3rem;
  text-align: center;
  border: none;
  outline: none;
  font-size: 16px;
  -webkit-appearance: textfield;
     -moz-appearance: textfield;
          appearance: textfield;
}
.product__container__detail__main__quantity__input input[type=number]::-webkit-outer-spin-button, .product__container__detail__main__quantity__input input[type=number]::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}
.product__container__detail__main__quantity__input input[type=number]::-moz-placeholder {
  color: #999;
}
.product__container__detail__main__quantity__input input[type=number]::placeholder {
  color: #999;
}
.product__container__detail__main__quantity__input button {
  background-color: #f0f0f0;
  border: none;
  padding: 0 10px;
  cursor: pointer;
  font-size: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 3rem;
  height: 100%;
}
.product__container__detail__main__quantity__input button:hover {
  background-color: #e0e0e0;
}
.product__container__detail__main__quantity__input button:focus {
  outline: none;
}
.product__container__detail__main__quantity__input-minus {
  color: #ff4d4d;
}
.product__container__detail__main__quantity__input-plus {
  color: #4caf50;
}
@media (max-width: 768px) {
  .product__container__detail__main__quantity__input {
    width: 120px;
    height: 35px;
  }
  .product__container__detail__main__quantity__input input[type=number] {
    width: 40px;
    font-size: 14px;
  }
  .product__container__detail__main__quantity__input button {
    width: 35px;
    font-size: 18px;
  }
}
.product__container__detail__main__button {
  display: flex;
  align-items: center;
  gap: 10px;
  margin-top: 20px;
}
.product__container__detail__main__button button {
  padding: 10px 20px;
  font-size: 1.25rem;
  border: none;
  border-radius: 8px;
  cursor: pointer;
  transition: background-color 0.3s ease;
}
.product__container__detail__main__button button:hover {
  background-color: #2c2c2c;
}
.product__container__detail__main__button button:focus {
  outline: none;
}
.product__container__detail__main__button-cart {
  background-color: #3d3d3d;
  color: white;
}
.product__container__detail__main__button-cart:hover {
  background-color: #4d4d4d;
}
.product__container__detail__main__button-wishlist {
  background-color: #ff4d4d;
  color: white;
}
.product__container__detail__main__button-wishlist:hover {
  background-color: #e63939;
}
.product__container__detail__main__button-wishlist.active {
  background-color: #ffa07a;
}
.product__container__detail__main__button-wishlist.active:hover {
  background-color: #ff7f50;
}
@media (max-width: 768px) {
  .product__container__detail__main__button {
    flex-direction: column;
    gap: 15px;
  }
  .product__container__detail__main__button button {
    width: 100%;
  }
}
.product__container__navigation {
  width: 100%;
}
.product__container__navigation__btn {
  width: 100%;
  display: flex;
  flex-direction: row;
}
.product__container__navigation__btn__empty {
  width: 100%;
  border-bottom: 2px solid #ccc;
}
.product__container__navigation button {
  width: 250px;
  color: #3d3d3d;
  border: none;
  border-bottom: 2px solid #ccc;
  outline: none;
  cursor: pointer;
  background-color: white;
  margin: 0;
}
.product__container__navigation .active {
  color: black;
  border-bottom: 2px solid #000;
}
.product__container__review {
  background-color: #f9f9f9;
  border-radius: 8px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
}
.product__container__review__container {
  background-color: white;
  border-radius: 8px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.05);
}
.product__container__review__container:not(:last-child) {
  margin-bottom: 20px;
}
.product__container__review__container__rating {
  margin-bottom: 20px;
}
.product__container__review__container__rating__title {
  font-size: 18px;
  font-weight: bold;
  margin-bottom: 10px;
}
.product__container__review__container__rating__container {
  display: flex;
  align-items: center;
  gap: 10px;
}
.product__container__review__container__rating__container__star {
  display: flex;
  gap: 0.5rem;
}
.product__container__review__container__rating__container__star span {
  cursor: pointer;
  transition: transform 0.3s ease;
  font-size: 14px;
}
.product__container__review__container__rating__container__star span:hover {
  transform: scale(1.2);
}
.product__container__review__container__rating__container__score {
  font-size: 18px;
  color: #555;
}
.product__container__review__container__text {
  margin-bottom: 20px;
}
.product__container__review__container__text__title {
  font-size: 18px;
  font-weight: bold;
  margin-bottom: 10px;
}
.product__container__review__container__text__area {
  width: 100%;
  padding: 10px;
  border: 1px solid #ddd;
  border-radius: 5px;
  font-size: 14px;
  resize: vertical;
  margin-bottom: 10px;
  transition: border-color 0.3s ease;
}
.product__container__review__container__text__area:focus {
  border-color: #999;
}
.product__container__review__container__text__button {
  padding: 10px 20px;
  background-color: #3d3d3d;
  color: white;
  border: none;
  border-radius: 5px;
  cursor: pointer;
  transition: background-color 0.3s ease;
}
.product__container__review__container__text__button:hover {
  background-color: #2c2c2c;
}
.product__container__review__container__all__list {
  margin-top: 20px;
}
.product__container__review__container__all__list__item {
  padding: 15px;
  background-color: #f8f8f8;
  border-radius: 8px;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.1);
  margin-bottom: 15px;
}
.product__container__review__container__all__list__item:last-child {
  margin-bottom: 0;
}
.product__container__review__container__all__list__item__info {
  display: flex;
  align-items: center;
  margin-bottom: 10px;
  gap: 2rem;
}
.product__container__review__container__all__list__item__info__star {
  display: flex;
  flex-direction: row;
  gap: 0.25rem;
}
.product__container__review__container__all__list__item__info strong {
  font-size: 16px;
  font-weight: 600;
}
.product__container__review__container__all__list__item__info span {
  font-size: 14px;
  color: #999;
}
.product__container__review__container__all__list__item__review p {
  font-size: 14px;
  color: #444;
  line-height: 1.5;
  margin: 0;
}
@media (max-width: 768px) {
  .product__container__review__container__all__list__item {
    padding: 10px;
  }
  .product__container__review__container__all__list__item__info {
    flex-direction: column;
    align-items: flex-start;
    gap: 5px;
  }
  .product__container__review__container__all__list__item__review p {
    font-size: 13px;
  }
}
.product__container__description {
  padding: 20px;
  background-color: #fff;
  border-radius: 8px;
  box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  margin-top: 20px;
}
.product__container__description h5 {
  font-size: 20px;
  font-weight: bold;
  margin-bottom: 10px;
}
.product__container__description p {
  font-size: 14px;
  line-height: 1.6;
  color: #333;
}
@media (max-width: 768px) {
  .product__container__review {
    padding: 15px;
  }
  .product__container__description {
    padding: 15px;
  }
  .product__container__description h5 {
    font-size: 18px;
  }
  .product__container__description p {
    font-size: 13px;
  }
}
.product__container__recommended__title {
  font-size: 24px;
  font-weight: bold;
  margin-top: 4rem;
}
.product__container__recommended__list {
  display: grid;
  grid-template-columns: repeat(10, 1fr);
  overflow-x: auto;
  gap: 1.25rem;
  padding: 2rem 0;
}
.product__container__recommended__list__item {
  position: relative;
  height: -moz-fit-content;
  height: fit-content;
  padding: 0.25rem;
  box-shadow: 0 4px 10px 3px rgba(0, 0, 0, 0.1882352941);
}
.product__container__recommended__list__item__image {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100px;
  margin-bottom: 1rem;
  cursor: pointer;
}
.product__container__recommended__list__item__image img {
  display: inline-block;
  text-align: center;
  height: 100%;
  -o-object-fit: contain;
     object-fit: contain;
}
.product__container__recommended__list__item button {
  background-color: white;
  color: #0047ff;
  width: 113px;
  height: 44px;
  border: 1px solid #0047ff;
  font-weight: bold;
  border-radius: 20px;
  transition: 0.3s all ease-in-out;
}
.product__container__recommended__list__item button:hover {
  background-color: #0047ff;
  color: white;
}
.product__container__recommended__list__item:hover .product__container__recommended__list__item__like {
  opacity: 1;
}
.product__container__recommended__list__item:hover .product__container__recommended__list__item__wishlist {
  opacity: 1;
}/*# sourceMappingURL=ProductPage.css.map */