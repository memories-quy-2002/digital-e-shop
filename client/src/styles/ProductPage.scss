.product {
    &__container {
        display: flex;
        flex-direction: column;
        padding: 3rem 15rem;
        gap: 4rem;
        &__detail {
            display: flex;
            flex-direction: row;
            justify-content: space-around;
            height: 750px;
            &__img {
                width: 500px;
                &__main {
                    img {
                        width: 100%;
                    }
                }
                &__sub {
                    margin-top: 1.5rem;
                    display: flex;
                    flex-direction: row;
                    justify-content: space-between;
                    img {
                        height: 100px;
                    }
                }
            }
            &__info {
                display: flex;
                flex-direction: column;
                justify-content: space-between;
            }
            &__quantity {
                height: 45px;
                display: flex;
                flex-direction: column;
                gap: 1rem;

                button {
                    width: 60px;
                    height: 47px;
                    padding: 0;
                    font-size: 24px;
                }
                input {
                    height: 47px;
                    width: 70px;
                    text-align: center;
                    font-size: 24px;
                    padding: 0;
                    border: 2px solid black;
                }
            }
            &__btn {
                display: flex;
                flex-direction: row;
                gap: 2rem;
                button {
                    width: 250px;
                    height: 80px;
                    font-size: 24px;
                    text-transform: uppercase;
                    border-radius: 20px;
                }
            }
        }
        &__navigation {
            width: 100%;
            &__btn {
                width: 100%;
                display: flex;
                flex-direction: row;
                &__empty {
                    width: 100%;
                    border-bottom: 2px solid #ccc;
                }
            }

            button {
                width: 250px;
                color: #3d3d3d;
                border: none;
                border-bottom: 2px solid #ccc;
                outline: none;
                cursor: pointer;
                background-color: white;
                margin: 0;
            }
            .active {
                color: black;
                border-bottom: 2px solid #000;
            }
        }
        &__description {
            text-align: justify;
            h5 {
                font-size: 28px;
                margin-bottom: 1.25rem;
            }
        }
        &__review {
            display: flex;
            flex-direction: column;
            gap: 1.75rem;
            h5 {
                font-size: 28px;
            }
            &__rating {
                strong {
                    font-size: 32px;
                }
                &__score {
                    display: flex;
                    flex-direction: row;
                    align-items: center;
                    gap: 1rem;
                }
            }
            &__text {
                display: flex;
                flex-direction: column;
                gap: 1rem;
                button {
                    width: 100px;
                }
                textarea {
                    padding: 1rem;
                    width: 40%;
                    margin-bottom: 1rem;
                }
            }
            &__list {
                h6 {
                    margin: 0;
                    font-size: 24px;
                }
            }
        }
        &__recommended {
            &__title {
                font-size: 24px;
                font-weight: bold;
                margin-top: 4rem;
            }
            &__list {
                display: grid;
                grid-template-columns: repeat(10, 1fr);
                overflow-x: auto;
                gap: 1rem;
                margin: 2rem 0;
                padding: 2rem 0;
                &__item {
                    position: relative;
                    padding: 1rem 0.5rem;
                    height: fit-content;
                    box-shadow: 0 4px 10px 3px rgba(0, 0, 0, 0.1882352941);
                    &__image {
                        display: flex;
                        justify-content: center;
                        align-items: center;
                        height: 120px;
                        margin-bottom: 1rem;
                        cursor: pointer;
                        img {
                            display: inline-block;
                            text-align: center;
                            height: 100%;
                            object-fit: contain;
                        }
                    }

                    button {
                        background-color: white;
                        color: #0047ff;
                        width: 113px;
                        height: 44px;
                        border: 1px solid #0047ff;
                        font-weight: bold;
                        border-radius: 20px;
                        transition: 0.3s all ease-in-out;
                        &:hover {
                            background-color: #0047ff;
                            color: white;
                        }
                    }
                    &:hover &__like {
                        opacity: 1;
                    }
                    &:hover &__wishlist {
                        opacity: 1;
                    }
                }
            }
        }
    }
}
