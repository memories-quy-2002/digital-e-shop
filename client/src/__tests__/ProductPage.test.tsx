import { render } from "@testing-library/react";
import ProductPage from "../components/pages/ProductPage";
import { MemoryRouter } from "react-router-dom";
import ToastProvider from "../context/ToastContext";

describe("ProductPage", () => {
    it("matches the ProductPage snapshot", async () => {
        const { asFragment } = render(
            <ToastProvider>
                <MemoryRouter>
                    <ProductPage />
                </MemoryRouter>
            </ToastProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
