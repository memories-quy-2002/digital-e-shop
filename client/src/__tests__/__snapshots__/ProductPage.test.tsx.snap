// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPage should match the ProductPage snapshot 1`] = `
<DocumentFragment>
  <main
    style="display: flex; align-items: center; flex-direction: column;"
  >
    <h1>
      404 Error
    </h1>
    <p>
      Page not found
    </p>
  </main>
  <div
    class="toast-container bottom-0 end-0 p-3"
    style="z-index: 1; position: fixed; bottom: 0px; right: 0px;"
  />
</DocumentFragment>
`;
